misc/multistream-select/src/dialer_select.rs

```diff
@@ -172,11 +190,123 @@ where
+
+                            let protocol = this.protocols.next().ok_or(NegotiationError::Failed)?;
+                            *this.state = SeqState::SendSimOpen { io, protocol: Some(protocol) };
+                        }
+                        Some(protocol) => {
+                            let p = Protocol::try_from(protocol.as_ref())?;
+                            if let Err(err) = Pin::new(&mut io).start_send(Message::Protocol(p.clone())) {
+                                return Poll::Ready(Err(From::from(err)));
+                            }
+                            log::debug!("Dialer: Proposed protocol: {}", p);
+
+                            *this.state = SeqState::FlushSimOpen { io, protocol }
+                        }
+                    }
+                }
+
+                SeqState::FlushSimOpen { mut io, protocol } => {
+                    match Pin::new(&mut io).poll_flush(cx)? {
+                        Poll::Ready(()) => {
+                            *this.state = SeqState::AwaitSimOpen { io, protocol }
+                        },
+                        Poll::Pending => {
+                            *this.state = SeqState::FlushSimOpen { io, protocol };
+                            return Poll::Pending
+                        },
+                    }
+                }
+
+                SeqState::AwaitSimOpen { mut io, protocol } => {
+                    let msg = match Pin::new(&mut io).poll_next(cx)? {
+                        Poll::Ready(Some(msg)) => msg,
+                        Poll::Pending => {
+                            *this.state = SeqState::AwaitSimOpen { io, protocol };
+                            return Poll::Pending
+                        }
+                        // Treat EOF error as [`NegotiationError::Failed`], not as
+                        // [`NegotiationError::ProtocolError`], allowing dropping or closing an I/O
+                        // stream as a permissible way to "gracefully" fail a negotiation.
+                        Poll::Ready(None) => return Poll::Ready(Err(NegotiationError::Failed)),
+                    };
+
+                    match msg {
+                        Message::Header(v) if v == HeaderLine::from(*this.version) => {
+                            *this.state = SeqState::AwaitSimOpen { io, protocol };
+                        }
+                        Message::Protocol(p) if p == SIM_OPEN_ID => {
+                            let selection = SimOpenPhase {
+                                state: SimOpenState::SendNonce{ io },
+                            };
+                            *this.state = SeqState::SimOpenPhase { selection, protocol };
+                        }
+                        Message::NotAvailable => {
+                            *this.state = SeqState::AwaitProtocol { io, protocol }
+                        }
+                        _ => return Poll::Ready(Err(ProtocolError::InvalidMessage.into()))
```

Comment:
so currently no fallback to old style in case the peer isn’t ready for V1SimOpen, right?

Reply:
I think it's supposed to be backwards compatible.

Reply:
When using `Version::V1SimOpen` the dialer expects either:

- `/libp2p/simultaneous-connect`  in which case the remote indicates that it operates in dialing mode as well, i.e. a simultaneously dialed connection or
- `na` in which case the remote indicates that it is a listener, i.e. not a simultaneously dialed connection, and thus we fall back to the "old style" via `SeqState::AwaitProtocol`.

Any other message would be a protocol violation, both with and without the simultaneous open extension.

@rkuhn does that make sense?

Reply:
Yes, that `na` bit was the missing puzzle piece, thanks!